### ISSUE TITLE
Comparison of two identical Algebraic instances hangs

### ISSUE THREAD

**Issue by rklaehn** [non/spire#415]
When comparing two Algebraic that should be exactly the same (but that do have different ASTs), the comparison hangs indefinitely, or at least much longer than I have patience for.

(-x, y) and (x, -y) are the local extrema of the polynomial. See http://www.wolframalpha.com/input/?i=x%C2%B3+-+x

By the way: I think that implementing #414 would provide a good way to write generic tests for Algebraic.

```scala
    val p = poly"x^3 - x"
    val pa = Polynomial.dense(p.coeffsArray.map(x ⇒ Algebraic(x)))
    val x = Algebraic(3).nroot(2).reciprocal()
    val y = Algebraic(2) / (Algebraic(3) * Algebraic(3).nroot(2))
    val y1 = pa(x)
    val y2 = pa(-x)
    require(-y.toBigDecimal === y1.toBigDecimal)
    require(y.toBigDecimal === y2.toBigDecimal)
    println((y - y2).signum)
    require(-y == y1)
    require(y == y2)
```

**Comment by tixxit** [comment]
I have an outstanding Algebraic refactor which fixes a number of bugs and speeds up performance. Ill check this against that branch when I have internet again. The examples you gave are pretty tame, so I'd suspect a bug (I do worse in the property tests in the redactor branch).

**Comment by rklaehn** [comment]
Not sure if this is the cause of the bug, but if you print y.toTreeString and y2.toTreeString, it seems that y2 is completely non-simplified:
```
y
 |/       
.^-.      
|  |*     
| .^.     
| | |sqrt 
2 3 3     
y2
                                                                                           |/                                 
                                              .--------------------------------------------^-------------.                    
                                              |+                                                         |                    
                         .--------------------^-------------.                                            |                    
                         |                                  |*                                           |                    
                         |                           .------^--------------------------.                 |                    
                         |*                          |                                 |*                |                    
               .---------^-------------.             |                       .---------^.                |                    
               |+                      |             |                       |+         |                |*                   
 .-------------^-.                     |             |         .-------------^-.        |         .------^-------------.      
 |*              |*                    |*            |         |*              |*       |         |                    |*     
.^------.       .^---.          .------^.            |        .^------.       .^---.    |         |             .------^.     
|       |*      |    |*         |*      |            |*       |       |*      |    |*   |         |*            |*      |     
| .-----^.      | .--^.   .-----^.      |      .-----^.       | .-----^.      | .--^.   |   .-----^.      .-----^.      |     
| |sqrt  |sqrt  | |-  |-  |sqrt  |sqrt  |sqrt  |sqrt  |sqrt   | |sqrt  |sqrt  | |-  |-  |-  |sqrt  |sqrt  |sqrt  |sqrt  |sqrt 
0 3      3      0 1   1   3      3      3      3      3      -1 3      3      1 1   1   1   3      3      3      3      3    
```

**Comment by tixxit** [comment]
Algebraic doesn't do any simplification, but with the refactor it also doesn't force all divisions to the top, so the tree will be better. In this case I'd say the required precision for that tree is just killing the time. Hoping the flight I'm about to get on has WiFi :)

**Comment by tixxit** [comment]
This is fixed in my [new-algebraic branch](https://github.com/tixxit/spire/tree/topic/new-algebraic). The 2 comparisons listed both return "immediately" with `true`. That branch also implements #414 (though for any size rational polynomials).

**Comment by tixxit** [comment]
This is fixed in #427
